graph/src/blockchain/block_stream.rs

```diff
@@ -1,15 +1,89 @@
 use anyhow::Error;
+use async_stream::stream;
 use futures03::Stream;
 use std::sync::Arc;
 use thiserror::Error;
+use tokio::sync::mpsc::{self, Receiver, Sender};
 
 use super::{Block, BlockPtr, Blockchain};
 use crate::components::store::BlockNumber;
 use crate::firehose;
 use crate::{prelude::*, prometheus::labels};
 
+pub struct BufferedBlockStream<C: Blockchain> {
+    inner: Pin<Box<dyn Stream<Item = Result<BlockStreamEvent<C>, Error>> + Send>>,
+}
+
+impl<C: Blockchain + 'static> BufferedBlockStream<C> {
+    pub fn spawn_from_stream(
+        stream: Box<dyn BlockStream<C>>,
+        size_hint: usize,
+    ) -> Box<dyn BlockStream<C>> {
+        let (sender, receiver) = mpsc::channel::<Result<BlockStreamEvent<C>, Error>>(size_hint);
+        crate::spawn(async move { BufferedBlockStream::stream_blocks(stream, sender).await });
+
+        Box::new(BufferedBlockStream::new(receiver))
+    }
+
+    pub fn new(mut receiver: Receiver<Result<BlockStreamEvent<C>, Error>>) -> Self {
```

Comment:
Doesn't need to be `pub`.

Reply:
I disagree, I think the whole point is to expose the functions in a way they can be used separate but I can remove it

Reply:
Ok if you want to keep it.